zebrad/tests/common/launch.rs

```diff
@@ -228,12 +235,61 @@ pub fn spawn_zebrad_for_rpc_without_initial_peers<P: ZebradTestDirExt + std::fmt
 
     let rpc_address = config.rpc.listen_addr.unwrap();
 
-    zebrad.expect_stdout_line_matches("activating mempool")?;
     zebrad.expect_stdout_line_matches(&format!("Opened RPC endpoint at {}", rpc_address))?;
 
     Ok((zebrad, rpc_address))
 }
 
+/// Wait for lightwalletd to sync to Zebra's tip.
+///
+/// "Adding block" and "Waiting for block" logs stop when `lightwalletd` reaches the tip.
+/// But if the logs just stop, we can't tell the difference between a hang and fully synced.
+/// So we assume `lightwalletd` will sync and log large groups of blocks,
+/// and check for logs with heights near the mainnet tip height.
+#[tracing::instrument]
+pub fn wait_for_zebrad_and_lightwalletd_tip<
+    P: ZebradTestDirExt + std::fmt::Debug + std::marker::Send + 'static,
+>(
+    mut lightwalletd: TestChild<TempDir>,
+    mut zebrad: TestChild<P>,
+    test_type: LightwalletdTestType,
+) -> Result<(TestChild<TempDir>, TestChild<P>)> {
+    let lightwalletd_thread = std::thread::spawn(move || -> Result<_> {
+        tracing::info!(?test_type, "waiting for lightwalletd to sync to the tip");
+
+        lightwalletd
+            .expect_stdout_line_matches(crate::common::sync::LIGHTWALLETD_SYNC_FINISHED_REGEX)?;
```

Comment:
I think we'll want to:

In the `zebrad` thread:
- wait for Zebra to sync to the tip
- keep checking the `zebrad` logs until `lightwalletd` syncs to the tip
- then finish the thread

In the `lightwalletd` thread:
- keep checking the `lightwalletd` logs until Zebra syncs to the tip
- then keep asking `lightwalletd` for its height via gRPC, until it is greater than or equal to the `zebrad` height
- then finish the thread

This probably needs a channel from the `zebrad` to the `lightwalletd` thread.

Reply:
Was the idea behind `wait_for_zebrad_tip: bool` to make this optional or should it always do this?

Reply:
Some tests need to wait until they sync to the public network chain tip. Others run disconnected from the network, so they just need to load whatever state they have.

And then lightwalletd will sync whatever zebrad has in both kinds of tests.